### PR TITLE
Add full ffmpeg and dependencies

### PR DESCRIPTION
**Pull request by maxrd2**
FFmpeg used in flatpak doesn't support required codecs and demuxers, which breaks some subtitle composer features.

### CODE REVIEW

**Comment by travier** [comment]
We're already using the ffmpeg-full extension: https://github.com/flathub/org.kde.subtitlecomposer/blob/master/org.kde.subtitlecomposer.json#L27. Is that not enough?

**Comment by maxrd2** [comment]
> We're already using the ffmpeg-full extension: https://github.com/flathub/org.kde.subtitlecomposer/blob/master/org.kde.subtitlecomposer.json#L27. Is that not enough?

For some reason full in that extension doesn't mean full.
It doesn't have support for a lot of codecs and demuxers/formats.

That ffmpeg is built by disabling everything but few common codecs and formats.

Here is related issue: https://invent.kde.org/multimedia/subtitlecomposer/-/issues/72

**Comment by travier** [comment]
That's a lot of new dependencies. @hfiguiere Do you know if this is the right approach?

**Comment by hfiguiere** [comment]
to me it's not you don't need frillions of .json. It makes everything hard to maintain.

**Comment by travier** [comment]
@maxrd2 Where did you copy those manifest from (if you did that) / how did you generate them?

**Comment by maxrd2** [comment]
I'm not sure if it's the right approach, am pretty new to flatpak universe :)

SubtitleComposer needs FFmpeg version with all decoders and demuxers compiled in - don't care about encoders right now, but might need them in the future.

If there is some place that it can be included from it's fine for me, although I doubt it - [kdenlive](https://github.com/flathub/org.kde.kdenlive/blob/master/org.kde.kdenlive.json) is also building/bundling it.

**Comment by maxrd2** [comment]
> @maxrd2 Where did you copy those manifest from (if you did that) / how did you generate them?

Have used json's from here: git@github.com:tinywrkb/flatpak-shared-modules.git, but have modified/updated most of them.

I can merge them into single JSON if you think that's easier/better.

**Comment by maxrd2** [comment]
@Erick555 I've removed everything that is not needed, so it doesn't look like kitchen sink, have added host access explanation to commit. Probably it some more will be needed in the future when more features are added to the application.

@travier Have merged all needed manifests into main json file, like other projects seem to do.

Will this work?

BTW I'd like to squash commits 863da49 a5329ee 1dd29a6 before merging this

**Comment by Erick555** [comment]
> SubtitleComposer needs FFmpeg version with all decoders and demuxers compiled in - don't care about encoders right now, but might need them in the future.

So x264 and x265 dependencies are added only for the future without current use?

File: org.kde.subtitlecomposer.json
Comment (Erick555) [inline org.kde.subtitlecomposer.json:404]: There is `"/lib/*.a",` global cleanup rule. Does this makes sense then?

File: org.kde.subtitlecomposer.json
Comment (Erick555) [inline org.kde.subtitlecomposer.json:434]: Same

File: org.kde.subtitlecomposer.json
Comment (maxrd2) [reply org.kde.subtitlecomposer.json:404]: i think libx264/5 are also used for decoding?

**Comment by maxrd2** [comment]
> > SubtitleComposer needs FFmpeg version with all decoders and demuxers compiled in - don't care about encoders right now, but might need them in the future.
> 
> So x264 and x265 dependencies (encoders) are added only for the future without current use?

i think libx264/5 are also used for decoding?

File: org.kde.subtitlecomposer.json
Comment (maxrd2) [reply org.kde.subtitlecomposer.json:404]: libx265_main10.a is 10bpp color x265 build, then there is 12bpp build, both are linked together with main 8bpp build into a final libx265.so.... if i'm not mistaken

**Comment by travier** [comment]
Let's add a README to explain the permissions like we did for Ark: https://github.com/flathub/org.kde.ark
In this case, you probably don't need host but home access

**Comment by maxrd2** [comment]
> Let's add a README to explain the permissions like we did for Ark: https://github.com/flathub/org.kde.ark

ok

> In this case, you probably don't need host but home access

That would exclude external usb drives, network shares and similar paths no?


BTW are all `x-checker-data` tags I've included in manifest recommended/required or optional? Those could possibly break builds along the way?

**Comment by travier** [comment]
> > In this case, you probably don't need host but home access
> 
> That would exclude external usb drives, network shares and similar paths no?

Good point. Let's keep host.

> BTW are all `x-checker-data` tags I've included in manifest recommended/required or optional? Those could possibly break builds along the way?

It's better to have them. If updates introduce issues, we can always disable the automatic update checker in Flathub.

**Comment by travier** [comment]
This mostly LGTM. Have all the subtitlecomposer patches been submitted upstream already?

**Comment by travier** [comment]
And I've just realized that you're the upstream maintainer 🙂

**Comment by maxrd2** [comment]
> This mostly LGTM. Have all the subtitlecomposer patches been submitted upstream already?

They are taken from upstream master branch and they will be included in 0.8.0 which I'll release soon-ish.

```
* 2ec9412d Fixed invalid memory read
* 3fe2fb49 Fixed crash when closing subtitle
* 17811ce5 Fix crashes on empty lines list
* fa7b2dd6 Removed QSharedData instances from stack
* 911b94b3 Fixed memory leaks
* 82f907f4 LinesItemDelegate: fix inline rich editor crash
* 4f4f560e Fix compilation with ffmpeg5 #63
* d8f9797d VideoPlayer: Fix usage of deprecated/removed AVCodec option #68
```

**Review by travier — APPROVED** [review]
Not tested but this LGTM

**Comment by travier** [comment]
You we wait for the release then to make this less verbose?

**Comment by maxrd2** [comment]
> Should we wait for the release then to make this less verbose?

I prefer to push this ASAP, as have realized few days ago after looking into vobsub bug report that flatpak subtitle composer version is very broken.

It will be at least few weeks before 0.8.0 is released (and wanted to do to it way sooner :disappointed:).

Will do another PR then and drop all the patches.

**Comment by Erick555** [comment]
I think there are still potentially unneeded dependencies and options in ffmpeg, i.e. svt-av1 is used for [encoding](https://github.com/FFmpeg/FFmpeg/blob/master/configure#L275). srt is for streaming. vmaf is some quality assessment, etc.

**Comment by Erick555** [comment]
Same thing for [amf](https://github.com/FFmpeg/FFmpeg/blob/master/configure#L334)

**Comment by maxrd2** [comment]
@Erick555 done

**Comment by Erick555** [comment]
I'm glad we are getting there :smile:

> amf-headers - AMD HW acceleration

As I noted before this is only **encoding** acceleration. There is no amf decoder in ffmpeg: https://github.com/FFmpeg/FFmpeg/blob/n4.2.1/configure#L3007-L3085

Above is confirmed in ffmpeg build [log](https://buildbot.flathub.org/api/v2/logs/262644/raw). No amf available:
```
Enabled hwaccels:
av1_nvdec               mjpeg_vaapi             vc1_vdpau
av1_vaapi               mpeg1_nvdec             vp8_nvdec
av1_vdpau               mpeg1_vdpau             vp8_vaapi
h263_vaapi              mpeg2_nvdec             vp9_nvdec
h264_nvdec              mpeg2_vaapi             vp9_vaapi
h264_vaapi              mpeg2_vdpau             vp9_vdpau
h264_vdpau              mpeg4_nvdec             wmv3_nvdec
hevc_nvdec              mpeg4_vaapi             wmv3_vaapi
hevc_vaapi              mpeg4_vdpau             wmv3_vdpau
hevc_vdpau              vc1_nvdec
mjpeg_nvdec             vc1_vaapi
```

As you disable all encoders the amf dependency becomes no-op unless you can prove otherwise.

> libsoxr - audio resampling

Is audio resampling somewhat relevant to subitlecomposer?

BTW: interesting thing happens here - libsoxr links to ffmpeg from runtime and app's ffmpeg links to libsoxr which means ffmpeg transiently links to different ffmpeg :smile::
```
/usr/lib/gcc/x86_64-unknown-linux-gnu/11.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: warning: libavcodec.so.58, needed by /app/lib/libsoxr.so, may conflict with libavcodec.so.59
/usr/lib/gcc/x86_64-unknown-linux-gnu/11.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: warning: libavutil.so.56, needed by /usr/lib/x86_64-linux-gnu/libavcodec.so.58, may conflict with libavutil.so.57
```

> vmaf - perceptual video quality assessment algorithms - video filters/possibly conversion

is this relevant for subtitlecomposer?

> zimg - video scaling/colorspace conversion

same question

**Comment by maxrd2** [comment]
> I'm glad we are getting there smile

I just hope when we get there it will be where I wanted to be in the first place :smiley: 

> > amf-headers - AMD HW acceleration
> 
> As I noted before this is only **encoding** acceleration. There is no amf decoder in ffmpeg: https://github.com/FFmpeg/FFmpeg/blob/master/configure#L3172-L3265

It's possible and so you say, but I can see references to AMF types in a bit of FFmpeg source code, including some decoders like flvdec and some other files that might be used by other decoders/filters and similar.

AMF are just headers, it doesn't directly add any libraries or binaries, so what exactly is the rationale for excluding it?

> Above is confirmed in ffmpeg build [log](https://buildbot.flathub.org/api/v2/logs/262644/raw). No amf available:
> 
> ```
> Enabled hwaccels:
> av1_nvdec               mjpeg_vaapi             vc1_vdpau
> av1_vaapi               mpeg1_nvdec             vp8_nvdec
> av1_vdpau               mpeg1_vdpau             vp8_vaapi
> h263_vaapi              mpeg2_nvdec             vp9_nvdec
> h264_nvdec              mpeg2_vaapi             vp9_vaapi
> h264_vaapi              mpeg2_vdpau             vp9_vdpau
> h264_vdpau              mpeg4_nvdec             wmv3_nvdec
> hevc_nvdec              mpeg4_vaapi             wmv3_vaapi
> hevc_vaapi              mpeg4_vdpau             wmv3_vdpau
> hevc_vdpau              vc1_nvdec
> mjpeg_nvdec             vc1_vaapi
> ```

I'm not sure if confirmed is very accurate in this context. Yes there is support for decoders but will they be in the best shape in which they can be if most dependencies are stripped away? Quick glance over the build logs for sure won't give very accurate insight into internal workings of huge set of algorithms that are implemented in FFmpeg.

I was talking with @herzenschein about ffmpeg issues when he started adding subtitlecomposer to flathub one year ago. Several versions have been tried, and most of them had issues decoding most common codecs and formats. For example H.264 videos playing was janky - a lot of videos had decoding speed/quality issues and some would not open at all. Finally he discovered `org.freedesktop.Platform.ffmpeg-full` which **seemed** to work fine.

> As you disable all encoders the amf dependency becomes no-op unless you can prove otherwise.

Look at libavformat/flvdec.c?

> > libsoxr - audio resampling
> 
> FFmpeg has [libswresample](https://ffmpeg.org/doxygen/2.3/group__lswr.html) which as I see is used by [subtitlecomposer](https://invent.kde.org/search?search=swresample&nav_source=navbar&project_id=730&group_id=1549&search_code=true). Why another resample library is needed and how it's being used?
> 
> BTW: interesting thing happens here - libsoxr links to ffmpeg from runtime and app's ffmpeg links to libsoxr which means ffmpeg transiently links to different ffmpeg smile:
> 
> ```
> /usr/lib/gcc/x86_64-unknown-linux-gnu/11.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: warning: libavcodec.so.58, needed by /app/lib/libsoxr.so, may conflict with libavcodec.so.59
> /usr/lib/gcc/x86_64-unknown-linux-gnu/11.3.0/../../../../x86_64-unknown-linux-gnu/bin/ld: warning: libavutil.so.56, needed by /usr/lib/x86_64-linux-gnu/libavcodec.so.58, may conflict with libavutil.so.57
> ```

Which means that libsoxr is not being built in clean evironment - env already contains libav libraries and libsoxr is linking against them and bunch of others:

libsoxr built in clean environment with `cmake ../soxr-${pkgver}-Source -DCMAKE_BUILD_TYPE='Release' -DCMAKE_INSTALL_PREFIX='/usr' -DBUILD_EXAMPLES='OFF' -DBUILD_SHARED_LIBS='ON' -DWITH_AVFFT='ON' -DWITH_LSR_BINDINGS='ON' -DWITH_OPENMP='ON' -DWITH_PFFFT='ON'` produces following with `ldd /usr/lib/libsoxr-lsr.so`
```
        linux-vdso.so.1 (0x00007ffc7352b000)
        libsoxr.so.0 => /usr/lib/libsoxr.so.0 (0x00007f38512c0000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f38510d9000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f3850ff1000)
        libgomp.so.1 => /usr/lib/libgomp.so.1 (0x00007f3850fa8000)
        /usr/lib64/ld-linux-x86-64.so.2 (0x00007f385139f000)
```

libsoxr built by flatpak-builder with `cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_EXAMPLES=OFF -DBUILD_SHARED_LIBS=ON -DWITH_AVFFT=ON -DWITH_LSR_BINDINGS=ON -DWITH_OPENMP=ON -DWITH_PFFFT=ON` produces this with `ldd /app/lib/libsoxr-lsr.so`:
```
        linux-vdso.so.1 (0x00007ffcb2ffb000)
        libsoxr.so.0 => /app/lib/libsoxr.so.0 (0x00007f7ec689a000)
        libc.so.6 => /usr/lib/x86_64-linux-gnu/libc.so.6 (0x00007f7ec66ba000)
        libavcodec.so.58 => /usr/lib/x86_64-linux-gnu/libavcodec.so.58 (0x00007f7ec6000000)
        libm.so.6 => /usr/lib/x86_64-linux-gnu/libm.so.6 (0x00007f7ec5eb7000)
        libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f7ec6674000)
        libgcc_s.so.1 => /usr/lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7ec665a000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f7ec6926000)
        libavutil.so.56 => /usr/lib/x86_64-linux-gnu/libavutil.so.56 (0x00007f7ec5c00000)
        libvpx.so.6 => /usr/lib/x86_64-linux-gnu/libvpx.so.6 (0x00007f7ec5800000)
        liblzma.so.5 => /usr/lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f7ec662e000)
        libdav1d.so.5 => /usr/lib/x86_64-linux-gnu/libdav1d.so.5 (0x00007f7ec567e000)
        libz.so.1 => /usr/lib/x86_64-linux-gnu/libz.so.1 (0x00007f7ec6615000)
        libaom.so.3 => /usr/lib/x86_64-linux-gnu/libaom.so.3 (0x00007f7ec4e00000)
        libfdk-aac.so.2 => /usr/lib/x86_64-linux-gnu/libfdk-aac.so.2 (0x00007f7ec5b44000)
        libmp3lame.so.0 => /usr/lib/x86_64-linux-gnu/libmp3lame.so.0 (0x00007f7ec5605000)
        libopus.so.0 => /usr/lib/x86_64-linux-gnu/libopus.so.0 (0x00007f7ec65b6000)
        libspeex.so.1 => /usr/lib/x86_64-linux-gnu/libspeex.so.1 (0x00007f7ec5b26000)
        libvorbis.so.0 => /usr/lib/x86_64-linux-gnu/libvorbis.so.0 (0x00007f7ec5af8000)
        libvorbisenc.so.2 => /usr/lib/x86_64-linux-gnu/libvorbisenc.so.2 (0x00007f7ec4d55000)
        libopenh264.so.6 => /usr/lib/x86_64-linux-gnu/openh264/extra/libopenh264.so.6 (0x00007f7ec4800000)
        libva.so.2 => /usr/lib/x86_64-linux-gnu/libva.so.2 (0x00007f7ec55d7000)
        libpthread.so.0 => /usr/lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7ec55b6000)
        libdl.so.2 => /usr/lib/x86_64-linux-gnu/libdl.so.2 (0x00007f7ec65ae000)
        libva-drm.so.2 => /usr/lib/x86_64-linux-gnu/libva-drm.so.2 (0x00007f7ec65a9000)
        libvdpau.so.1 => /usr/lib/x86_64-linux-gnu/libvdpau.so.1 (0x00007f7ec5eb1000)
        libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007f7ec4c0d000)
        libvulkan.so.1 => /usr/lib/x86_64-linux-gnu/libvulkan.so.1 (0x00007f7ec4bac000)
        libogg.so.0 => /usr/lib/x86_64-linux-gnu/libogg.so.0 (0x00007f7ec5ea6000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7ec4400000)
        libdrm.so.2 => /usr/lib/x86_64-linux-gnu/libdrm.so.2 (0x00007f7ec55a0000)
        libXext.so.6 => /usr/lib/x86_64-linux-gnu/libXext.so.6 (0x00007f7ec558b000)
        libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007f7ec4b80000)
        libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007f7ec5585000)
```

> I guess you may drop `-DWITH_AVFFT=ON` in order to prevent this from happening, see https://github.com/chirlu/soxr/blob/master/CMakeLists.txt#L69-L70 . Dropping whole module if it's unneeded may make more sense though.

I guess I might drop AVFFT support, which might stand for use libav's (FFmpeg's) fast fourier transforms or not., and it might make all algorithms depending on FFT (a lot of them) slower or not.

Audio resampling quality will directly affect quality of speech recognition inside subtitlecomposer.

> > vmaf - perceptual video quality assessment algorithms - video filters/possibly conversion
> 
> is this relevant for subtitlecomposer?

Am I using it directly in SC for something? No.
Is FFmpeg using it for something that might make SC composer quality or performance better? it's possible

> > zimg - video scaling/colorspace conversion
> 
> same question

I believe good video scaling and colorspace conversion is needed and used yes, if that's what you're asking me.

It took a bit of time and effort to have precise and [good quality picture](https://subtitlecomposer.kde.org/2020/10/18/colorspaces-and-conversions.html) it would be shame for all of it to go to waste because of one less dependency.

Multiple ffmpeg versions being used in flatpak is concerning and seems like a big can of worms.

Sincerely I prefer altogether removing Subtitle Composer flatpak from flathub (and KDE discover) then having people installing broken binary distribution and dismissing it as a bad/broken software without even bothering on filing a bug report.

**Comment by maxrd2** [comment]
Is it possible and doable to bundle AppImage binary inside flatpak without depending on org.kde.Platform?

AppImage [here](https://subtitlecomposer.kde.org/download.html) includes latest/stable/working Qt5/KF5/FFmpeg/pocketsphinx and everything else needed and it's only 56MB in size compared to current post-stripping 70-80MB flatpak with 300+MB of org.kde.Platform dependency

**Comment by maxrd2** [comment]
libsoxr with PFFFT instead of AVFFT seem to solve the .so dependency problem

**Comment by Erick555** [comment]
> It's possible and so you say, but I can see references to AMF types in a bit of FFmpeg source code, including some decoders like flvdec and some other files that might be used by other decoders/filters and similar.

Do you mean like [those](https://github.com/FFmpeg/FFmpeg/commit/6cac3a3b0812e945d7a5d42f5bf5e99332df34e7#diff-428035afc741ba4cfa5903bf1b361727bca66157efd807408966ab25a283210dR94) introduced in 2006 while **Advanced Media Framework** was published in 2016?

> AMF are just headers, it doesn't directly add any libraries or binaries, so what exactly is the rationale for excluding it?

Aren't you getting it backwards? You are were putting there something like 20 new dependencies copy pasted from somewhere without any kind of proof they are related to this app and expect I will prove they are useless one by one (which I'm doing). Isn't it you who is supposed to provide exact reason for including it?

> was talking with @herzenschein about ffmpeg issues when he started adding subtitlecomposer to flathub one year ago. Several versions have been tried, and most of them had issues decoding most common codecs and formats. For example H.264 videos playing was janky - a lot of videos had decoding speed/quality issues and some would not open at all.

That's right yet half of your changes weren't related to this breakage. The X is broken and you did Y but it doesn't fix X so why Y it's there?

> Audio resampling quality will directly affect quality of speech recognition inside subtitlecomposer.

The question was if SC really uses libsoxr for resampling or just the default ffmpeg library. Adding library won't magically improve quality when it's not being called.

> Am I using it directly in SC for something? No.
> Is FFmpeg using it for something that might make SC composer quality or performance better? it's possible

It's more possible it doesn't. Why you insist on doing something you don't know?

> I believe good video scaling and colorspace conversion is needed and used yes, if that's what you're asking me.

I asked you if SC uses zimg/zscale filter. It's not question of faith but fact.

> Sincerely I prefer altogether removing Subtitle Composer flatpak from flathub (and KDE discover) then having people installing broken binary distribution and dismissing it as a bad/broken software without even bothering on filing a bug report.

I agree fixing SC on flathub is important issue but the fact is this PR was overly complicated for no reason. Maintaining 20 new dependencies would waste much time.

**Comment by herzenschein** [comment]
So, if I understand correctly, with your additions to ffmpeg @maxrd2 you don't see those decoding issues anymore, yes? And these are the same settings used for the AppImage's ffmpeg?

Is there a way to test whether those decoding issues happen?

(aside from the obvious "manually looking at the videos being reproduced in the app")

To be completely honest, all this ffmpeg thing is way above my head, but give me some way to test and I shall.

If we want to ship this ASAP, then I imagine the practical approaches would be:

* We can test NOW to see if each codec/library/whatever makes a difference
* We can ship it as is, and remove the unnecessary things later (hopefully @Erick555 would still help us with this issue :smile: )

So to summarize, the "unnecessary candidates" to test would be the following, correct?

* svt-av1
* srt
* vmaf
* amf
* zimg

**Comment by Erick555** [comment]
The svt-av1 and srt were already removed. What I reccomend is to drop the rest questionable things and merge this ASAP. That would fix most if not all underlying issues.

After that you can make new PR that adds some new library which could be tested. If it does any difference then nobody would argue against adding it.

**Comment by maxrd2** [comment]
> Do you mean like [those](https://github.com/FFmpeg/FFmpeg/commit/6cac3a3b0812e945d7a5d42f5bf5e99332df34e7#diff-428035afc741ba4cfa5903bf1b361727bca66157efd807408966ab25a283210dR94) introduced in 2006 while Advanced Media Framework support was introduced in ffmpeg in [2017](https://git.videolan.org/?p=ffmpeg.git;a=commitdiff;h=9ea6607d294526688ab1b1342cb36ee159683e88)?

Yes... that's what I meant. Have done a quick grep of the sources, and made bad assumption. My bad.

I really don't want to spend time to analyze ffmpeg sources to figure how each and every of zillion decoders, demuxers, hardare support library performs on multitude of systems. Have been using ffmpeg API for years, and am pretty happy with decision that upstream makes.

Some of the FFmpeg defaults are decided on runtime based on system, and some are changed upstream.
For example your mention of soxr and swr resampling filter and how it won't magically enable itself unless I change it from Subtitle Composer isn't entirely true from my point of view.

Yes swr is the default [right now](https://github.com/FFmpeg/FFmpeg/blob/master/libswresample/options.c#L112) and has been for some time. And there were people that wanted to remove [soxr](https://ffmpeg.org/pipermail/ffmpeg-devel/2017-March/207927.html). But the thing is that swr is default because that AVOption.default_val says 0 (SWR_ENGINE_SWR), but that could change upstream to SWR_ENGINE_SOXR or to something else because upstream discovered that something else works better.

> As I noted before this is only encoding acceleration. There is no amf decoder in ffmpeg: https://github.com/FFmpeg/FFmpeg/blob/master/configure#L3172-L3265

No there isn't AMF decoder at the moment, and from that link it would seems there is also no libopenh264 decoder, which is not true - it also adds decoder which is not very obvious/well documented. That same decoder when included at one point in past ended up handling all H,264 by default and was causing all kinds of problems I have mentioned before.

So my reasoning is - don't exclude/remove dependencies that tomorrow could become default for similar reasons, especially with flathub autobot updating and creating PR every time there is upstream release. This would mean I or someone else would have to re-test a ton of things manually every time some upstream dependency changes. OTOH I trust that upstream FFmpeg won't (intentionally) change defaults to something worse.

> Aren't you getting it backwards? You were putting there something like 20 new dependencies copy pasted from somewhere without any kind of proof they are related to this app and expect I will prove they are useless one by one (which I'm doing). Isn't it you who is supposed to provide exact reason for including it?

I didn't intend to keep first commits that included 100 manifests... some of them unused. I was expecting someone more experienced in flatpak universe will suggest some along the lines of "hey why don't you use instead extension org.bim-bum-bam.Platform.ffmpeg-that-just-works-for-everyone instead of building your 100 things" - that didn't happen and here we are analyzing ffmpeg sources and discussing how upstream is handling encoders/decoders/filters and defaults.

There are other 22 binary distributions that subtitlecomposer is being [built for](https://build.opensuse.org/project/monitor/home:maxrd2) across several architectures, and for each of them I trust upstream distro to provide a sane working package of ffmpeg and other dependencies, in rare cases where there is a problem (fedora, appimage) the dependency (FFmpeg, Qt, KF5, pocketsphinx) is built per upstream defaults/recommendation or in similar manner that other sane distributions do.

Regarding your questions about soxr/zimg/zscale and other filters, no I'm not explicitly selecting them at the moment, I do whatever ffmpeg chooses to select by default, and I'd like to keep them there in case FFmpeg "decides" to select something else per default or per specific case.

So i have removed all the dependencies for which I'm 100% sure that are not and will not be used. What is left if ffmpeg, 1 one needed dependency and 8 for which I'm not 100% sure that are irrelevant.

**Comment by maxrd2** [comment]
@Erick555 I removed all questionable things, if I remove anything else the build will break.

**Comment by Erick555** [comment]
> @Erick555 I removed all questionable things, if I remove anything else the build will break.

You removed also things I didn't question but if final result is satisfactory for you then it's fine

> Have been using ffmpeg API for years, and am pretty happy with decision that upstream makes.

The thing is if you want to follow upstream decisions/defaults then you don't need to pass tons of enable/disable options by yourself. The defaults are something that is handled automatically without involvement from someone who builds it. For example omitting --enable-libsoxr is enough to stop building it you don't need to additionally pass --disable-libsoxr. This is really that simple. That's why I said you over complicated this. Every thing we discussed here was some non-default change that you made.

What you wanted is to pass few --enable-xxx for stuff that you need but isn't enabled by upstream decision.

> No there isn't AMF decoder at the moment, and from that link it would seems there is also no libopenh264 decoder, which is not true

libopenh264 decoder defined in [the same place](https://github.com/FFmpeg/FFmpeg/blob/master/configure#L3366-L3367) below.

**Comment by maxrd2** [comment]
Having 5 or 8 less features doesn't change much if you're (not) affected by them.
Also have placed a note on website in case of problems to use other type of distribution.
It's better than it was - at least all basic things work now more or less.

Thank you for your time and patience @Erick555 @travier @herzenschein

**Review by travier — APPROVED** [review]
LGTM

File: org.kde.subtitlecomposer.json
Comment (maxrd2) [reply org.kde.subtitlecomposer.json:9]: yes... 

32d693999862c17d123b609d16589551812a8551 [Add full filesystem access permission](https://github.com/flathub/org.kde.subtitlecomposer/pull/10/commits/32d693999862c17d123b609d16589551812a8551) 

Desktop portal permits access only to user selected file, application
needs to load media files next to subtitle file for automatic video laod
to work. Also bitmap subtitles are stored in multiple files and user
selects only one.

Unsure if user script management requires it.

Comment (maxrd2) [reply org.kde.subtitlecomposer.json:13]: Have cleaned up manifest in [512ad1f](https://github.com/flathub/org.kde.subtitlecomposer/pull/10/commits/512ad1f9f05d6e29a6fa0be81a34ea6254d5448e)

**Comment by maxrd2** [comment]
Seems kross build is broken... there is 404 when download version 5.99 and 5.100 requires KF5/ecm 5.100?

Have added patches for that and removed it.

**Comment by herzenschein** [comment]
Build 7283 seems to be working fine. No apparent issues.

**Comment by maxrd2** [comment]
Is everybody ok with this?
Should I just press the merge button?

**Pull request merged**
